src/becca/entities/brevision.ts

```diff
@@ -1,68 +1,76 @@
 "use strict";
 
-const protectedSessionService = require('../../services/protected_session.js');
-const utils = require('../../services/utils.js');
-const dateUtils = require('../../services/date_utils.js');
-const becca = require('../becca.js');
-const AbstractBeccaEntity = require('./abstract_becca_entity.js');
-const sql = require('../../services/sql.js');
-const BAttachment = require('./battachment.js');
+import protectedSessionService = require('../../services/protected_session');
+import utils = require('../../services/utils');
+import dateUtils = require('../../services/date_utils');
+import becca = require('../becca');
+import AbstractBeccaEntity = require('./abstract_becca_entity');
+import sql = require('../../services/sql');
+import BAttachment = require('./battachment');
+import { AttachmentRow, RevisionRow } from './rows';
+
+interface ContentOpts {
+    /** will also save this BRevision entity */
+    forceSave?: boolean;
+}
+
+interface GetByIdOpts {
+    includeContentLength?: boolean;
+}
 
 /**
  * Revision represents a snapshot of note's title and content at some point in the past.
  * It's used for seamless note versioning.
- *
- * @extends AbstractBeccaEntity
  */
-class BRevision extends AbstractBeccaEntity {
+class BRevision extends AbstractBeccaEntity<BRevision> {
     static get entityName() { return "revisions"; }
     static get primaryKeyName() { return "revisionId"; }
     static get hashedProperties() { return ["revisionId", "noteId", "title", "isProtected", "dateLastEdited", "dateCreated",
                                             "utcDateLastEdited", "utcDateCreated", "utcDateModified", "blobId"]; }
 
-    constructor(row, titleDecrypted = false) {
+    revisionId?: string;
+    noteId: string;
+    type: string;
+    mime: string;
+    isProtected: boolean;
+    title: string;
+    blobId?: string;
```

Comment:
Why can this and some of the others be undefined

Reply:
`blobId` is optional because in `BNote#saveRevision` we create a `BRevision` without a `blobId`.

The approach I've had to establish whether fields are mandatory or optional is to make all fields mandatory by default and make them optional whenever it was omitted in uses of the constructor/method/field.